src/crypto/slow-hash.c

```diff
@@ -198,6 +198,21 @@ extern int aesb_pseudo_round(const uint8_t *in, uint8_t *out, const uint8_t *exp
     } while (0)
 #endif
 
+#define VARIANT2_2_1() \
+  do if (variant >= 2) \
+  { \
+    *U64(hp_state + (j ^ 0x10)) ^= hi; \
+    *(U64(hp_state + (j ^ 0x10)) + 1) ^= lo; \
+  } while (0); \
+
+#define VARIANT2_2_2() \
+  do if (variant >= 2) \
+  { \
+    hi ^= *U64(hp_state + (j ^ 0x20)); \
+    lo ^= *(U64(hp_state + (j ^ 0x20)) + 1); \
```

Comment:
The idea to use XOR to break fused MUL+ADD implementation is great, but do we need to use scratchpad for that? We have plenty of data readily available in registers here - result from previous AES, division_result, sqrt_result etc. It will be much smaller hit for CPU if we don't do XOR with memory location. XOR will still break fused MUL+ADD as long as it's done with some non-constant value.

Reply:
Although, division and sqrt results are not yet available here because of high calculation latency. But everything else is available and can be used for XOR.

Reply:
The simple `hi ^= something1` and `lo ^= something2` will be enough to break fused MUL+ADD. Let's keep this tweak simple, current version seems to be over complicated without any visible benefit to outweigh this complexity.

Reply:
> The idea to use XOR to break fused MUL+ADD implementation is great, but do we need to use scratchpad for that? We have plenty of data readily available in registers here - result from previous AES, division_result, sqrt_result etc. It will be much smaller hit for CPU if we don't do XOR with memory location. XOR will still break fused MUL+ADD as long as it's done with some non-constant value.

All of things mentioned in this list should be easier to "tweak" to anyone "sitting" on a design for CNv2 since they are already required for the computation of the hi/lo value. I will inspect a bit further though.

> The simple `hi ^= something1` and `lo ^= something2` will be enough to break fused MUL+ADD. Let's keep this tweak simple, current version seems to be over complicated without any visible benefit to outweigh this complexity.

This tweak **is very simple**. However, it is reducing some computations that can be done in parallel. I don't see how this type of parallelism is a magic feat that only a full CPU/GPU can do ... ?

Reply:
>  I don't see how this type of parallelism is a magic feat that only a full CPU/GPU can do ... ?

You understand it the wrong way. This type of parallelism (with shuffles) is **the only** one CPU/GPU can do, ASIC are capable of much more. So this is the only way how CPU/GPU can be competitive.

P.S. Do you know how original Cryptonight ASICs achieved such high hashrates? They reduced the critical path to 20-27 bits needed to calculate indices in the scratchpad and did everything else in parallel in the background. Major parts of AES and MUL weren't needed to progress the main loop, they could be done with lag because they were only needed when the same 16-byte chunk was hit again. No CPU/GPU can do that.

Reply:
>> I don't see how this type of parallelism is a magic feat that only a full CPU/GPU can do ... ?
>
> You understand it the wrong way. This type of parallelism (with shuffles) is the only one CPU/GPU can do, ASIC are capable of much more. So this is the only way how CPU/GPU can be competitive.

This seems extremely close to deflecting from my argument (because an ASIC can still outperform either way). I _think_ what you are getting is that the CPU already has more available fetch capabilities from L3/memory (and so do many GPUs), and so this is increasing the relative cost of production of an ASIC whereas its fixed for CPU/GPU. Something like that?

> P.S. Do you know how original Cryptonight ASICs achieved such high hashrates? They reduced the critical path to 20-27 bits needed to calculate indices in the scratchpad and did everything else in parallel in the background. Major parts of AES and MUL weren't needed to progress the main loop, they could be done with lag because they were only needed when the same 16-byte chunk was hit again. No CPU/GPU can do that.

I don't see how this was possible. Selecting an address depends on AES or mul/add results from the immediate prior step. Unless there is some shortcut through the AES stage or something ... ?

The ASIC could delay/queue writes, but you seem to be describing something different.

Reply:
>I think what you are getting is that the CPU already has more available fetch capabilities from L3/memory (and so do many GPUs), and so this is increasing the relative cost of production of an ASIC whereas its fixed for CPU/GPU. Something like that?

Yes, pretty close. Please comment on my proposed change to VARIANT2_2 (scroll down the page to see my last comments).

Reply:
> I don't see how this was possible

Me neither, I got "enlightened" by FPGA developers - they use this technique in their 22 KH/s CNv1 (current Monero PoW) bitstream for FPGA.